File: 11-Binary-Search/java/07-BookAllocationProblem.java
class Solution {
    public int findPages(int[] nums, int m) {
        int n = nums.length;

        if (m > n) {
            return -1;
        }

        int low = Arrays.stream(nums).max().getAsInt();
        int high = Arrays.stream(nums).sum();

        while (low <= high) {
            int mid = low + (high - low) / 2;

            if (countStudents(nums, mid) <= m) {
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }

        return low;
    }

    private int countStudents(int[] nums, int maxPages) {
        int students = 1;
        int pagesAssigned = 0;

        for (int pages : nums) {
            if (pagesAssigned + pages > maxPages) {
                students++;
                pagesAssigned = pages;
            } else {
                pagesAssigned += pages;
            }
        }

        return students;
    }
}

File: 11-Binary-Search/python/07_book_allocation_problem.py
class Solution:
    def findPages(self, nums: List[int], m: int) -> int:
        n = len(nums)
        if m > n:
            return -1

        def count_students(max_pages):
            students = 1
            pages_assigned = 0
            for pages in nums:
                if pages_assigned + pages > max_pages:
                    students += 1
                    pages_assigned = pages
                else:
                    pages_assigned += pages
            return students

        low = max(nums)
        high = sum(nums)

        while low <= high:
            mid = (low + high) // 2
            if count_students(mid) <= m:
                high = mid - 1
            else:
                low = mid + 1

        return low
